src/api/helpers.jl

```diff
@@ -52,35 +52,52 @@ end
 # emulating it with the less desirable form of creating closure handles directly in
 # `@cfunction` with `$f`.
 # This helper translates between the two preferred forms for each respective language.
-function h5a_iterate_helper(loc_id::hid_t, attr_name::Ptr{Cchar}, ainfo::Ptr{H5A_info_t}, @nospecialize(f::Any))::herr_t
-    return f(loc_id, attr_name, ainfo)
+function h5a_iterate_helper(loc_id::hid_t, attr_name::Ptr{Cchar}, ainfo::Ptr{H5A_info_t}, @nospecialize(data::Any))::herr_t
+    f, err_ref = data
+    try
+        return herr_t(f(loc_id, attr_name, ainfo))
+    catch err
+        err_ref[] = err
+        return herr_t(-1)
+    end
 end
 
 """
     h5a_iterate(f, loc_id, idx_type, order, idx = 0) -> hsize_t
 
-Executes [`h5a_iterate`](@ref h5a_iterate(::hid_t, ::Cint, ::Cint, ::Ptr{hsize_t}, ::Ptr{Cvoid}, ::Ptr{Cvoid}))
-with the user-provided callback function `f`, returning the index where iteration ends.
+Executes [`h5a_iterate`](@ref h5a_iterate(::hid_t, ::Cint, ::Cint,
+::Ptr{hsize_t}, ::Ptr{Cvoid}, ::Ptr{Cvoid})) with the user-provided callback
+function `f`, returning the index where iteration ends.
 
 The callback function must correspond to the signature
 ```
-    f(loc::HDF5.API.hid_t, name::Ptr{Cchar}, info::Ptr{HDF5.API.H5A_info_t}) -> HDF5.API.herr_t
+f(loc::HDF5.API.hid_t, name::Ptr{Cchar}, info::Ptr{HDF5.API.H5A_info_t}) -> Union{Bool, Integer}
 ```
-where a negative return value halts iteration abnormally, a positive value halts iteration
-successfully, and zero continues iteration.
+where a negative return value halts iteration abnormally (triggering an error),
+a `true` or a positive value halts iteration successfully, and `false` or zero
+continues iteration.
 
 # Examples
 ```julia-repl
 julia> HDF5.API.h5a_iterate(obj, HDF5.API.H5_INDEX_NAME, HDF5.API.H5_ITER_INC) do loc, name, info
            println(unsafe_string(name))
-           return HDF5.API.herr_t(0)
+           return false
        end
 ```
 """
 function h5a_iterate(@nospecialize(f), obj_id, idx_type, order, idx = 0)
+    err_ref = Ref{Any}(nothing)
```

Comment:
Why can't the reference always be an integer and instead of the `isnothing` checks, the value `jlerr < 0` is instead done?

Reply:
`jlerr` is the error object that is caught in the helper function, not an integer.

Reply:
I see, thanks for the clarification.